Improve Encryption at Rest documentation

ndragazis opened the pull request:
This PR introduces a major rewrite of the EaR document. The initial motivation for this PR was to fully cover all our supported key providers with working examples, and to add instructions for key rotation. However, many other improvements were made along the way.

Main changes in this PR:
* Add a high-level description for every key provider. Mention limitations.
* Better organize existing provider-specific instructions by placing them into clearly separated, tabbed sections.
* Add instructions for the replicated key provider. Mention explicitly that it cannot be used as default option for user or system encryption, and that it does not support key rotation.
* Add more examples for KMS and GCP to cover all credential types.
* Document missing configuration options.
* Add a new section for key rotation.

Notes:
* Some of the patches in this series have been cherry-picked from Laszlo's wip branch.
* This PR is expected to conflict with the Azure Key Vault PR, which should be merged first. (https://github.com/scylladb/scylladb/pull/23920/)
* Support for KMIP system keys in the Replicated Key Provider is currently broken. (https://github.com/scylladb/scylladb/issues/24443)

Fixes scylladb/scylla-enterprise#3535.
Refs scylladb/scylla-enterprise#3183.

Only doc changes. No backport is needed.

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/24558)

**Changed Files:**
- [operating-scylla/security/encryption-at-rest](http://previews.docs.scylladb.com/scylladb/24558/operating-scylla/security/encryption-at-rest)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

elcallio commented on docs/operating-scylla/security/encryption-at-rest.rst:
```diff
@@ -5,71 +5,103 @@ Encryption at Rest
-Encryption at Rest works at table level granularity, so you can choose to encrypt only sensitive tables. For both system and table data, you can use different algorithms that are supported by `OpenSSL <https://www.openssl.org/>`_ in a file block encryption scheme. 
 
-.. note:: SSTables of a particular table can have different encryption keys, use different encryption algorithms, or not be encrypted at all - at the same time. 
+Encryption at Rest works at table granularity, so you can choose to encrypt only sensitive tables. For both system and table data, you can use different block
```
Perhaps phrase as "can work at table granularity"?

ndragazis replied:
But there is no other granularity for user data. Only the level of configuration can vary.

elcallio replied:
True, but semantically here, I think the user cares more about the granularity of the configuration. So "can be configured at table gran." maybe?

elcallio commented on docs/operating-scylla/security/encryption-at-rest.rst:
```diff
@@ -5,71 +5,103 @@ Encryption at Rest
-As SSTables are immutable, tables are encrypted only once, as a result of memtable flush, compaction, or upgrade (with :doc:`Nodetool upgradesstables </operating-scylla/nodetool-commands/upgradesstables>`).
+As SSTables are immutable, they are encrypted only once, as a result of memtable flush, compaction, or upgrade (with :doc:`Nodetool upgradesstables
+</operating-scylla/nodetool-commands/upgradesstables>`). Similarly, system artifacts are encrypted when they are written. Encryption configuration takes effect
+at the time of write, and is persisted in the given SSTable's metadata (``Scylla.db``), or the given system artifact's (e.g. commit log's) config file,
```
Very nit, but could we keep these files to ~80 char width? Even in a fairly wide browser window, seeing the diff here I get loads of forced line breaks which makes reading rather stuttered.

ndragazis replied:
Nit and time consuming, but ok.

elcallio commented on docs/operating-scylla/security/encryption-at-rest.rst:
```diff
@@ -5,71 +5,103 @@ Encryption at Rest
+mode of operation) from the SSTable's metadata (``Scylla.db``). ScyllaDB extracts the key from the key provider (which is recorded in the same SSTable metadata
+file), and then decrypts the other SSTable components with the key.
 
+When ScyllaDB reads an encrypted system artifact, such as a commit log, it fetches the algorithm descriptor and the key provider from the commit log's config
```
Should probably phrase as "commitlog segment metadata file" or something. Config sounds like a singular entity, like scylla.yaml, which is not true.

elcallio commented on docs/operating-scylla/security/encryption-at-rest.rst:
```diff
@@ -81,206 +113,336 @@ When encrypting the system tables or SSTables, you need to state which provider
+     - Stores table keys in a ScyllaDB table where the table itself is encrypted using the system key
+   * - KMIP Key Provider
+     - ``KmipKeyProviderFactory``
+     - External key management server
```
... using KMIP (Key Management Interop Protocol) server protocol?

ndragazis replied:
Ack, it's properly phrased in the KMIP subsection. I'll fix it in the table as well.

elcallio commented on docs/operating-scylla/security/encryption-at-rest.rst:
```diff
@@ -81,206 +113,336 @@ When encrypting the system tables or SSTables, you need to state which provider
+   does not require any external resources. In production environments, it is recommended to use an external KMS instead.
 
-You should also consider keeping the key directory on a network drive (using TLS for the file sharing) to avoid having keys and data on the same storage media, in case your storage is stolen or discarded. 
+The Local Key Provider is the default key provider for ScyllaDB. It stores the encryption keys locally on disk in a text file. The location of this file is
```
The replicated provider is default, no?

ndragazis replied:
The Local is the default in `scylla.yaml`. The Replicated is the default in the table schema. I need to make it more clear.

elcallio commented on docs/operating-scylla/security/encryption-at-rest.rst:
```diff
@@ -81,206 +113,336 @@ When encrypting the system tables or SSTables, you need to state which provider
+The user has the option to generate the key(s) themselves, or let ScyllaDB generate the key(s) for them. When creating a key or locating an existing key, the
+key provider extracts the raw key material from the server and stores it in a loading cache in memory.
+
+Key rotation is supported, but the rotation must be performed by the user.
```
... on the KMIP server level. Or something like that maybe.

elcallio commented on docs/operating-scylla/security/encryption-at-rest.rst:
```diff
@@ -81,206 +113,336 @@ When encrypting the system tables or SSTables, you need to state which provider
+<ear-cipher-algorithms>`.
+
+The KMS Key Provider maintains data encryption keys in a loading cache to reuse them for the encryption of multiple artifacts. Once a key expires from the
+cache, it generates a new one from the KMS.
```
More "via" the KMS.

ndragazis replied:
Ack.

elcallio commented on docs/operating-scylla/security/encryption-at-rest.rst:
```diff
@@ -304,52 +466,107 @@ If you are using AWS KMS to encrypt tables or system information, add the KMS in
+
+   .. note::
+
+      Note that either ``endpoint``, ``aws_region`` or ``aws_use_ec2_region`` must be set (one of them is required for connection).
```
Should we also include that scylla will default to check the standard AWS env vars for ID/key + try load .credentials?

elcallio commented on docs/operating-scylla/security/encryption-at-rest.rst:
```diff
@@ -369,454 +586,1115 @@ If you are using Google GCP KMS to encrypt tables or system information, add the
+                  gcp_project_id: myproject
+                  gcp_location: global
+                  gcp_credentials_file: /home/user/.config/gcloud/application_default_credentials.json
+                  master_key: mykeyring/mykey
```
Same thing here. Scylla will try to read the gcloud defaults.

elcallio commented on docs/operating-scylla/security/encryption-at-rest.rst:
```diff
@@ -369,454 +586,1115 @@ If you are using Google GCP KMS to encrypt tables or system information, add the
+   ;
+
+.. note::
+   Existing SSTables will be automatically encrypted during the next compaction. To encrypt them immediately instead, run
```
Technically, it is not "existing" sstables that apply. Maybe "existing sstables will not be modified, but data will apply the new encryption configuration during next compaction" or similar?

ndragazis replied:
How about: "Existing SSTables will remain unencrypted until the next compaction re-creates them using the new encryption configuration.".

ndragazis replied:
It's hard to strike a balance between accuracy and simplicity.

elcallio replied:
Works for me.

elcallio reviewed (commented):
A few small nits and questions. Overall looks good. Someone else should read through this as well to make sure it makes sense for the end user.

ndragazis commented:
v2 Changelog:
* s/Encryption at Rest works at table granularity/Encryption at Rest can be configured at table granularity
* Rewrap at 80 chars (instead of 160), except for some bullet lists which are kept at 160 for readability.
* s/commit log config/commit log segment metadata file
* Fix table description for KMIP key provider.
* Clarify that the local key provider is the default in `scylla.yaml` and the replicated key provider is the default in the table schema.
* Clarify that key rotation for KMIP must be done manually at the server.
*  Add notes for credential resolution in KMS and GCP hosts.

ndragazis commented:
v3 Changelog:
* Rewrite the description for the `system_key_file` option in the Replicated Key Provider. List the two possible values in bullets. Easier to read.
* Fix the default value for `system_key_directory`. It is `<system_key_directory>/system/<key_name>`, not `<system_key_directory>/<key_name>`.
* Add the `key_name` and `system_key_directory` options in `user_info_encryption` as well. They work the same way as in `system_info_encryption`.

xemul commented:
@Annamikhlin , please review

ndragazis commented on docs/operating-scylla/security/encryption-at-rest.rst:
```diff
@@ -5,284 +5,547 @@ Encryption at Rest
+  select different configuration properties for each data class.
 
-.. _Replicated:
+* Encryption for user data can also be configured through the
```
I think that presenting the options in a list would be even better:
```markdown
* Encryption for user data can be configured at two levels:

  - Node-level configuration: Using the ``user_info_encryption`` section in
    ``scylla.yaml``.

  - Table-level configuration: Using the ``scylla_encryption_options`` in the
    table schema.

  The table-level options are both finer-grained and coarser-grained than the
  SSTable encryption defaults in ``scylla.yaml``:

  - They are finer-grained because they override any possible defaults in
    ``scylla.yaml`` for a particular table.

  - They are coarser-grained because they apply to all nodes in the cluster that
    write out new SSTables for the table, while the options configured in
    ``scylla.yaml`` apply only to a particular node.
```

ndragazis commented:
I know. I'm working on it.

ndragazis commented:
v4 Changelog:
* Rebased and resolved conflicts.
* Added tabs for Azure.
* Added missing commas in commands in the "Encrypt a New Table" section.
* Removed the invalid statement that the KMS's `GenerateDataKey` API generates only 128 and 256-bit keys.
* Removed references to GCP key aliases. There are no aliases in GCP.
* Marked Azure's `master_key` config option as required instead of optional. Technically, it's optional, but probably nobody cares, and it's marked as required for AWS and GCP anyway.

ndragazis commented:
@scylladb/scylla-maint please consider merging so that it makes it into 2025.4.
